streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -1836,6 +1841,33 @@ public void maybeThrowTaskExceptionsFromProcessingThreads() {
         }
     }
 
+    // track the size of the transaction buffer on each iteration to predict when it will be exceeded in advance
+    private long lastUncommittedBytes = 0L;
+
+    boolean needsCommit(final boolean updateDelta) {
+        if (maxUncommittedStateBytes < 0) {
+            // if our transaction buffers are unbounded, we never need to force an early commit
+            return false;
+        }
+
+        // force an early commit if the uncommitted bytes exceeds or is *likely to exceed* the configured threshold
+        final long uncommittedBytes = tasks.approximateUncommittedStateBytes();
+
+        final long deltaBytes = Math.max(0, uncommittedBytes - lastUncommittedBytes);
+
+        final boolean needsCommit =  maxUncommittedStateBytes > -1 && uncommittedBytes + deltaBytes > maxUncommittedStateBytes;
```

Comment:
Do we really need `maxUncommittedStateBytes > -1`? Isn't this always true once we pass line 1851?

Reply:
Yep! I did some refactoring and didn't spot the dead code :smile: